### PR TITLE
First Draft: cookietemple.cfg linter

### PR DESCRIPTION
**Pull request by Imipenem**

### CODE REVIEW

File: cookietemple/lint/template_linter.py [hunk 1]
```diff
@@ -583,3 +591,93 @@ def init_changelog_lint_content(self) -> list:
...
+        # if there were any missing sections, let linter fail
+        if missing_sections:
+            miss_section_info = 'Cookietemple config file misses section' + 's' if len(missing_sections) > 1 else ''
+            self.linter_ctx.failed.append(('general-7', miss_section_info + f': {" ".join(section for section in missing_sections)}'))
+            return False
+        else:
+            self.linter_ctx.passed.append(('general-7', 'All required cookietemple.cfg sections were found!'))
+            return True
+
+    def check_section(self, section_items, section_name) -> bool:
```
Comment (Imipenem) [inline cookietemple/lint/template_linter.py:651]: @Zethson Im not quite happy with those longer if statements and the function design, as their functionality is very similar. You have any ideas on redesigning?

File: cookietemple/lint/template_linter.py [hunk 2]
```diff
@@ -583,3 +591,93 @@ def init_changelog_lint_content(self) -> list:
...
+            self.linter_ctx.failed.append(('general-7', 'Aborted config section linting. Fix missing sections first!'))
+
+    def check_missing_sections(self, parsed_sections) -> bool:
+        """
+        Examine cookietemple config file for missing sections
+        :param parsed_sections: All parsed sections from the user's cookietemple.cfg file
+        """
+        sections = ['bumpversion', 'bumpversion_files_whitelisted', 'bumpversion_files_blacklisted', 'sync_files_blacklisted', 'sync_level']
+        missing_sections = []
+        # for every section check whether its in the parsed sections or not
```
Comment (Zethson) [inline cookietemple/lint/template_linter.py:638]: it is

do not use abbreviations like it's and its is even grammatically wrong.

File: cookietemple/lint/template_linter.py [hunk 3]
```diff
@@ -583,3 +591,93 @@ def init_changelog_lint_content(self) -> list:
...
+        for section in sections:
+            if section not in parsed_sections:
+                missing_sections.append(section)
+        # if there were any missing sections, let linter fail
+        if missing_sections:
+            miss_section_info = 'Cookietemple config file misses section' + 's' if len(missing_sections) > 1 else ''
+            self.linter_ctx.failed.append(('general-7', miss_section_info + f': {" ".join(section for section in missing_sections)}'))
```
Comment (Zethson) [inline cookietemple/lint/template_linter.py:645]: is `f'miss_section_info{section for section in missing sections}'` not possible?

File: cookietemple/lint/template_linter.py (hunk 2 above)
Comment (Imipenem) [reply cookietemple/lint/template_linter.py:638]: will fix

File: cookietemple/lint/template_linter.py (hunk 1 above)
Comment (Zethson) [reply cookietemple/lint/template_linter.py:651]: Maybe it is possible to use an Ordered inserted dict and then just go through every pair and check whether it is correctly available?
-> `{bumpversion: section0: (current_version, linting-code), section1: (another_version, linting-code)}`

If any of those fail we can just build a message from that

We know which key (= function) failed and we know the exact failure and an error code.
You could e.g. use `general-7-1` and `general-7-2`

File: cookietemple/lint/template_linter.py (hunk 3 above)
Comment (Zethson) [reply cookietemple/lint/template_linter.py:645]: Sorry, yes.

File: cookietemple/info/info.py [hunk 4]
```diff
@@ -133,7 +133,7 @@ def output_table(templates_to_print: list, handle: str) -> None:
         table.add_column("Version", justify="left")
 
         for template in templates_to_print:
-            table.add_row(f'[bold]{template[0]}', template[1], template[2], template[3], template[4])
+            table.add_row(f'[bold]{template[0]}', template[1], f'{template[2]}\n', template[3], template[4])
```
Comment (Zethson) [inline cookietemple/info/info.py:136]: why?

File: cookietemple/list/list.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def list_available_templates(self) -> None:
         table.add_column("Version", justify="left")
 
         for template in templates_to_tabulate:
-            table.add_row(f'[bold]{template[0]}', template[1], template[2], template[3], template[4])
+            table.add_row(f'[bold]{template[0]}', template[1], f'{template[2]}\n', template[3], template[4])
```
Comment (Zethson) [inline cookietemple/list/list.py:61]: why?

File: cookietemple/info/info.py (hunk 4 above)
Comment (Imipenem) [reply cookietemple/info/info.py:136]: I added a blank line because otherwise it was hardly visible to distinguish between two templates

Comment (Zethson) [reply cookietemple/info/info.py:136]: Mhm, I guess I'll have to see this...

File: cookietemple/lint/template_linter.py [hunk 6]
```diff
@@ -583,3 +593,103 @@ def init_changelog_lint_content(self) -> list:
...
+        :param section_items: A pair (name, value) for all items in a section
+        :param section_name: The sections name
+        :param main_linter: The linter calling the function
+        """
+        # a set containig the section name as a key and its linting rules as values
+        # linting rules made of:
+        #   1.) a tuple with a variable_name and its value ('*' if value does not care for linting)
+        #   => NOTE: if one variable can have multiple valid values, they are separated by '|'
+        #
+        #   2.) a number, stating the minimum number of items in a section (-1 if number does not care)
```
Comment (Zethson) [inline cookietemple/lint/template_linter.py:665]: >  (-1 if number does not care)

-1 if number does not matter

File: cookietemple/lint/template_linter.py [hunk 7]
```diff
@@ -583,3 +593,103 @@ def init_changelog_lint_content(self) -> list:
...
+        check_set = {
+            'bumpversion': [('current_version', '*'), 1],
+            'bumpversion_files_whitelisted': [('dot_cookietemple', '.cookietemple.yml'), -1],
+            'sync_level': [('ct_sync_level', 'patch|minor|major'), 1],
+            'sync_files_blacklisted': [('changelog', 'CHANGELOG.rst'), -1]
+        }
+        return ConfigLinter.apply_section_linting_rules(section_name, section_items, check_set.get(section_name), main_linter)
+
+    @staticmethod
+    def apply_section_linting_rules(section: str, section_items, section_lint_rules, main_linter: TemplateLinter) -> bool:
```
Comment (Zethson) [inline cookietemple/lint/template_linter.py:675]: Not sure whether "apply" really is the correct word here.

Apply always sounds like doing something which is supposed to last.

We're just checking something here.

**Review by Zethson — APPROVED** [review]
Code scales, but is pretty ugly, which is not your fault.

Don't know whether this is good code design, but let's roll with it.

LGTM

**Pull request merged**